File: hto/16/hto_16.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:  return 1
        if n == 1:  return x
        if n == -1: return 1 / x
        half = self.myPow(x, n >> 1)
        rest = self.myPow(x, n % 2)
        return rest * half * half


if __name__ == "__main__":
    s = Solution()
    print(s.myPow(2.0, 10))

File: hto/16/hto_16.go
func myPow(x float64, n int) float64 {
    if n < 0 {
        return 1.0 / fastPow(x, -n)
    }
    return fastPow(x, n)
}

func fastPow(x float64, n int) float64 {
    if n == 0 {
        return 1.0
    }
    half := fastPow(x, n / 2)
    if n % 2 == 1 {
        return half * half * x
    }
    return half * half
}

func myPow(x float64, n int) float64 {
    if n == 0 {
        return 1
    } 
    if n == 1 {
        return x
    }
    if n == -1 {
        return 1.0 / x
    }
    half := myPow(x, n >> 1)
    rest := myPow(x, n & 1)
    return half * half * rest
}
